src/arch/src/aarch64/fdt.rs

```diff
@@ -300,27 +309,109 @@ fn generate_prop64(cells: &[u64]) -> Vec<u8> {
     append_begin_node(fdt, "cpus")?;
     // As per documentation, on ARM v8 64-bit systems value should be set to 2.
     append_property_u32(fdt, "#address-cells", 0x02)?;
     append_property_u32(fdt, "#size-cells", 0x0)?;
-
     let num_cpus = vcpu_mpidr.len();
     for (cpu_index, mpidr) in vcpu_mpidr.iter().enumerate() {
         let cpu_name = format!("cpu@{:x}", cpu_index);
         append_begin_node(fdt, &cpu_name)?;
         append_property_string(fdt, "device_type", "cpu")?;
         append_property_string(fdt, "compatible", "arm,arm-v8")?;
-        if num_cpus > 1 {
-            // If the microVM has more than 1 vcpu we need to enable the power
-            // state coordination interface (PSCI) which will decide for us which
-            // vcpu is running or halted.
-            append_property_string(fdt, "enable-method", "psci")?;
-        }
+        // The power state coordination interface (PSCI) needs to be enabled for
+        // all vcpus.
+        append_property_string(fdt, "enable-method", "psci")?;
         // Set the field to first 24 bits of the MPIDR - Multiprocessor Affinity Register.
         // See http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0488c/BABHBJCI.html.
         append_property_u64(fdt, "reg", mpidr & 0x7F_FFFF)?;
+
+        for cache in l1_caches.iter() {
+            // Please check out
+            // https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf,
+            // section 3.8.
+            append_property_u32(fdt, cache.type_.of_cache_size(), cache.size_ as u32)?;
+            append_property_u32(
+                fdt,
+                cache.type_.of_cache_line_size(),
+                cache.line_size as u32,
+            )?;
+            append_property_u32(
+                fdt,
+                cache.type_.of_cache_sets(),
+                cache.number_of_sets as u32,
+            )?;
+        }
+
+        // Some of the non-l1 caches can be shared amongst CPUs. You can see an example of a shared scenario
+        // in https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf,
+        // 3.8.1 Example.
+        let mut prev_level = 1;
+        let mut node = false;
+        for cache in non_l1_caches.iter() {
+            // We append the next-level-cache node (the node that specifies the cache hierarchy)
+            // in the next iteration. For example,
+            // L2-cache {
+            //      cache-size = <0x8000> ----> first iteration
+            //      next-level-cache = <&l3-cache> ---> second iteration
+            // }
+            // The cpus per unit cannot be 0 since the sysfs will also include the current cpu
+            // in the list of shared cpus so it need to be at least 1. Firecracker trusts the host.
+            assert!(cache.cpus_per_unit > 0);
```

Comment:
Would it make sense to return an error here ?

Reply:
Yup, it would be more elegant. I will return an error.